### PR TITLE
Implement loadable with lazy and suspense for next dynamic

### LINKED ISSUE #43147: [appDir] Dynamically importing a Client Component using the 'use client' directive from a Server Component causes "Error: object is not a function"
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

    Operating System:
      Platform: darwin
      Arch: arm64
      Version: Darwin Kernel Version 22.1.0: Sun Oct  9 20:14:30 PDT 2022; root:xnu-8792.41.9~2/RELEASE_ARM64_T8103
    Binaries:
      Node: 14.18.0
      npm: 6.14.15
      Yarn: 1.22.17
      pnpm: 7.14.1
    Relevant packages:
      next: 13.0.5-canary.3
      eslint-config-next: 13.0.5-canary.3
      react: 18.2.0
      react-dom: 18.2.0

### What browser are you using? (if relevant)

Firefox 107.0

### How are you deploying your application? (if relevant)

Vercel

### Describe the Bug

Dynamically importing with `next/dynamic` a Client Component that uses a `use/client` directive will cause an `Error: object is not a function` error in development. The production build runs fine though. 

Also noticed that adding `ssr: false` won't prevent the component from getting prerendered when running `next build`. Adding `suspense: true` or `suspense: false` won't make a difference either.

### Expected Behavior

We should be able to dynamically import a component that makes heavy use of JavaScript in the client from a Server Component and use a custom Suspense-based fallback.

### Link to reproduction - Issues with a link to complete (but minimal) reproduction code will be addressed faster

https://github.com/brvnonascimento/use-client-dynamic-bug

### To Reproduce

1. Run `yarn create next-app --experimental-app`
2. Create a Client Component by adding the `'use client'` directive
3. Use `next/dynamic` to import the given component
4. See `Error: object is not a function` in development

```tsx
import dynamic from "next/dynamic";
import { Suspense } from "react";

const ClientTest = dynamic(() => import("./ClientTest"));

export default function Home() {
  return (
    <div>
      <Suspense fallback={<h1>Loading...</h1>}>
        <ClientTest />
      </Suspense>
    </div>
  );
}
```

https://user-images.githubusercontent.com/34252418/202890528-83f01938-555a-426a-ab6f-4c3cb8885ab5.mov

### PR DESCRIPTION
**Pull request by huozhi**
### Summary

Migrate `next/dynamic` to implementation based on `React.lazy` and `Suspense`. Then it becomes easier to migrate the existing code in pages to layouts. Then we can support both `ssr` and `loading` option for `next/dynamic`.

For `loading` option, it will work like `Suspense`'s `fallback` property

```js
<Suspense fallback={loading}>
  <DynamicComponent />
 </Suspense>
```

For `ssr` option, by default `React.lazy` supports SSR, but we'll disable the `ssr: false` case for dynamic import in server components since there's no client side involved.

Then we don't need `suspense` option anymore as react >= 18 is always required. Mark it as deprecated.

It also supports to load client component dynamically in server components now.

#### Code code changes
* switch loadable component to `lazy` + `Suspense`
* will make sure it's retuning a module from `loader()` to `loader().then(mod => ({ default: mod.default || mod }))` since `lazy()` only accepts loader returning a module
* Inside suspense boundary, throwing an error for ssr: false, catch the error on server and client side and ignore it.
* Ignore options like ssr: false for server components since they're on server, doesn't make sense
* Remove legacy dynamic related transform
#### Feature changes
* `next/dynamic` will work in the same way across the board (appDir and pages)
* For the throwing error, will make it become a API that throws error later in the future, so users can customize more with `Suspense`
* You can load client components now in server components with dynamic. Resolves #43147

#### Tests
* existing dynamic tests all work
* add case: import client component and load through next/dynamic in server components

### CODE REVIEW

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 9d626a06304c8a2847f802310465706503139d76

`pnpm testheadless test/integration/next-dynamic-with-suspense/test/index.test.ts` 
- next/dynamic with suspense > should render server-side
- next/dynamic with suspense > should render client-side

<details>
<summary>Expand output</summary>

  ● next/dynamic with suspense › should render server-side

    expect(received).toContain(expected) // indexOf

    Expected substring: "Thing"
    Received string:    "<!DOCTYPE html><html><head><style data-next-hide-fouc=\"true\">body{display:none}</style><noscript data-next-hide-fouc=\"true\"><style>body{display:block}</style></noscript><meta charSet=\"utf-8\"/><meta name=\"viewport\" content=\"width=device-width\"/><meta name=\"next-head-count\" content=\"2\"/><noscript data-n-css=\"\"></noscript><script defer=\"\" nomodule=\"\" src=\"/_next/static/chunks/polyfills.js?ts=1667836754344\"></script><script src=\"/_next/static/chunks/webpack.js?ts=1667836754344\" defer=\"\"></script><script src=\"/_next/static/chunks/main.js?ts=1667836754344\" defer=\"\"></script><script src=\"/_next/static/chunks/pages/_app.js?ts=1667836754344\" defer=\"\"></script><script src=\"/_next/static/chunks/pages/index.js?ts=1667836754344\" defer=\"\"></script><script src=\"/_next/static/development/_buildManifest.js?ts=1667836754344\" defer=\"\"></script><script src=\"/_next/static/development/_ssgManifest.js?ts=1667836754344\" defer=\"\"></script><noscript id=\"__next_css__DO_NOT_USE__\"></noscript></head><body><div id=\"__next\"><div><p>Next.js Example</p>Loading...</div></div><script src=\"/_next/static/chunks/react-refresh.js?ts=1667836754344\"></script><script id=\"__NEXT_DATA__\" type=\"application/json\">{\"props\":{\"pageProps\":{}},\"page\":\"/\",\"query\":{},\"buildId\":\"development\",\"nextExport\":true,\"autoExport\":true,\"isFallback\":false,\"scriptLoader\":[]}</script></body></html>"

      25 |     const html = await renderViaHTTP(appPort, '/')
      26 |     expect(html).toContain('Next.js Example')
    > 27 |     expect(html).toContain('Thing')
         |                  ^
      28 |   })
      29 |
      30 |   it('should render client-side', async () => {

      at Object.<anonymous> (integration/next-dynamic-with-suspense/test/index.test.ts:27:18)

  ● next/dynamic with suspense › should render client-side

    expect(received).toBe(expected) // Object.is equality

    Expected: false
    Received: true

      32 |     const warnings = (await browser.log()).map((log) => log.message).join('\n')
      33 |
    > 34 |     expect(await hasRedbox(browser)).toBe(false)
         |                                      ^
      35 |     expect(warnings).toMatch(
      36 |       /"ssr: false" is ignored by next\/dynamic because you can not enable "suspense" while disabling "ssr" at the same time/gim
      37 |     )

      at Object.<anonymous> (integration/next-dynamic-with-suspense/test/index.test.ts:34:38)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Review by SukkaW — CHANGES_REQUESTED** [review]
Also, the test case added in #39676 should also be removed.

File: packages/next/shared/lib/dynamic.tsx [hunk 1]
```diff
@@ -41,22 +44,32 @@ export type LoadableFn<P = {}> = (
...
   delete loadableOptions.webpack
   delete loadableOptions.modules
 
-  // This check is necessary to prevent react-loadable from initializing on the server
-  if (!isServerSide) {
-    return LoadableInitializer(loadableOptions)
-  }
+  const NoSSRComponent = React.lazy(
+    (isServerSide
+      ? async () => ({ default: () => null })
```
Comment (sebmarkbage) [inline packages/next/shared/lib/dynamic.tsx:56]: For noSSR, the equivalent is not to return null but to throw an error. That will put the Suspense boundary in a client-rendered state on the server.

So this should be something like `() => { throw new Error('This lazy component is intentionally client rendered only.') }`.

File: packages/next/shared/lib/dynamic.tsx [hunk 2]
```diff
@@ -24,11 +23,21 @@ export type DynamicOptionsLoadingProps = {
...
+  return { default: mod.default || mod }
+}
+
 export type DynamicOptions<P = {}> = LoadableGeneratedOptions & {
   loading?: (loadingProps: DynamicOptionsLoadingProps) => JSX.Element | null
   loader?: Loader<P> | LoaderMap
   loadableGenerated?: LoadableGeneratedOptions
   ssr?: boolean
+  /**
+   * @deprecated `suspense` prop is not required any more
```
Comment (shuding) [inline packages/next/shared/lib/dynamic.tsx:39]:
```suggestion
   * @deprecated The `suspense` prop is not required anymore
```

File: packages/next/client/app-index.tsx [hunk 3]
```diff
@@ -113,6 +114,12 @@ if (document.readyState === 'loading') {
   DOMContentLoaded()
 }
 
+// Skip certain custom errors which are not expected to throw on client
+function onRecoverableError(err: any) {
+  if (err.digest === NEXT_DYNAMIC_NO_SSR_CODE) return
+  throw err
```
Comment (timneutkens) [inline packages/next/client/app-index.tsx:120]: What will happen when this case is hit? Will it mess with the stacktrace?

File: packages/next/export/worker.ts [hunk 4]
```diff
@@ -420,14 +422,20 @@ export default async function exportPage({
...
+          renderResult = await renderMethod(
+            req,
+            res,
+            page,
+            query,
+            // @ts-ignore
+            curRenderOpts
+          )
+        } catch (err: any) {
+          if (err.digest !== NEXT_DYNAMIC_NO_SSR_CODE) {
```
Comment (timneutkens) [inline packages/next/export/worker.ts:435]: It seems `NEXT_DYNAMIC_NO_SSR_CODE` fully overlaps with something @hanneslund has been working on to bail out of SSR, would be good to leverage that instead: https://github.com/vercel/next.js/pull/43603/files#diff-5becfaf098605c339c89c458fa973f977de17d9176b801680a04a92e91b340cbR35

Comment (huozhi) [reply packages/next/export/worker.ts:435]: but that seems not being handled on client? like for `onRecoverableError`

File: packages/next/client/app-index.tsx (hunk 3 above)
Comment (huozhi) [reply packages/next/client/app-index.tsx:120]: we shouldn't throw here, will change to the default one in react, it only reports or logs the error

File: packages/next/export/worker.ts (hunk 4 above)
Comment (timneutkens) [reply packages/next/export/worker.ts:435]: Probably should be!

Comment (huozhi) [reply packages/next/export/worker.ts:435]: yeah I think I can try to merge them together later!

**Review by shuding — APPROVED** [review]

**Comment by klarstrup** [comment]
This changed the undocumented way to programmatically preload `LoadableComponent`s from `dynamic`, it is now `LoadableComponent.preload()` rather than `LoadableComponent.render.preload()`.

**Comment by arishoham** [comment]
> This changed the undocumented way to programmatically preload `LoadableComponent`s from `dynamic`, it is now `LoadableComponent.preload()` rather than `LoadableComponent.render.preload()`.

@klarstrup looks like `.preload()` no longer exists on `LoadableComponent`. We are using `jest-next-dynamic` which seems to have broken with this PR, any chance you know how to preload a component now?

**Comment by klarstrup** [comment]
@arishoham as described the method is still there except it is now on the parent rather than `render`

**Comment by agusterodin** [comment]
As of Next 13.1 it appears that preload method doesn't exist as described above. It doesn't even appear to be an object.

Is there a known fix?

image

**Comment by agusterodin** [comment]
I see that there is future planned "offscreen" support in React. Will this a solution? 

https://reactjs.org/blog/2022/03/29/react-v18.html

**Comment by klarstrup** [comment]
> As of Next 13.1 it appears that preload method doesn't exist as described above. It doesn't even appear to be an object.
> 
> Is there a known fix?
> 
> image

It's still there as described above
image

**Comment by agusterodin** [comment]
Hmm. I get a different result when console printing. Am I printing this out properly?

image

**Comment by agusterodin** [comment]
It appears that this change that makes it no longer an object occurred on Next 13.0.7. If I console log in 13.0.6, I get an object with preload function as expected. Can you please clear lockfile and pin to latest version of Next to verify?

**Comment by davidgolden** [comment]
Sorry to butt in on your discussion, but I finally nailed down a minimal reproducible example of an issue I'm seeing that appears related to this change. This is the error itself: `Error: This Suspense boundary received an update before it finished hydrating. This caused the boundary to switch to client rendering. The usual way to fix this is to wrap the original update in startTransition.`.

It can be reproduced easily enough with this component which I believe should work fine:
```
import dynamic from "next/dynamic";
import {Component} from "react";
const SomeComponent = dynamic(() => import("../src/SomeComponent"));

export default class Index extends Component<any, any> {
    state = {
        test: "123",
    }

    componentDidMount() {
        this.setState({
            test: "234",
        })
    }

    render() {
        return <SomeComponent />
    }
}
```
Where `SomeComponent` can be anything you want. It doesn't like setting state inside `componentDidMount` when rendering a dynamic component. I'm on next@13.1.1.

**Comment by klarstrup** [comment]
@agusterodin did you try calling/logging `Timeline.preload` yet? You can use `console.dir` to see any properties of a function instead of the stringified function as `console.log`provides

**Comment by wadehammes** [comment]
@huozhi are there plans to fix the underlying issue here as stated above?

**Comment by arishoham** [comment]
@agusterodin @klarstrup the issue is the `preload` function no longer exists when `{ssr: false}`

**Comment by huozhi** [comment]
@davidgolden you should use `React.startTransition` for the setState in componentDidMount as the react warning suggested

**Comment by huozhi** [comment]
Adding the `preload()` back in https://github.com/vercel/next.js/pull/44832

**Comment by davidgolden** [comment]
> @davidgolden you should use `React.startTransition` for the setState in componentDidMount as the react warning suggested

This fixes it. I also just found that `useLayoutEffect` inside a dynamically loaded component causes the same issue. This feels like a bit of a breaking change that I'm surprised popped up in a minor version.

**Comment by icekhab** [comment]
@huozhi We can add "startTransition" to own code but it's problem when we use some library :)

**Comment by huozhi** [comment]
@icekhab could you file a new issue with more detailed repro and description for your case? Thanks

**Pull request merged**